### PR TITLE
Create HLD for programmatically sorting columns

### PR DESCRIPTION
**Pull request by mollykreis**
# Pull Request

## 🤨 Rationale

Create HLD for #874 

## 👩‍💻 Implementation

N/A

## 🧪 Testing

N/A

## ✅ Checklist

- [ ] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+In some cases, a client may want to configure their table to be sorted by a column that is not visible within the table. To enable this, a column within the table can be marked as hidden. A hidden column exists in the table to provide metadata that can be used for operations, such as sorting, but it is not rendered within the table.
+
+A column within the table may not have a one-to-one mapping to a field within the table's data. For example, a hyperlink column might use two fields from the data: a string to use as the hyperlink's display and a string to use as the hyperlink's href. Because of this, a column must define which data field is used when sorting that column. That field is specified by a column definition by implementing an `operationalDataFieldName` getter that returns the name of the field to use for sorting. For example, the text column would return the value of `fieldName` and the hyperlink column would return the field name associated with the display string.
```
Comment (mollykreis) [inline packages/nimble-components/src/table/specs/table-column-sort-hld.md:30]: Note: I'm not sold on the name `operationalDataFieldName`, but I struggled to come up with the right name for this property. I'm open to suggestions that others have.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+#### Summary of new attributes on the base table column
+
+| attribute name | type                                                                                                         | default value                   | description                                                     |
+| -------------- | ------------------------------------------------------------------------------------------------------------ | ------------------------------- | --------------------------------------------------------------- |
+| sort-direction | `TableColumnSortDirection`, defined as `{none: undefined, ascending: 'ascending', descending: 'descending'}` | `TableColumnSortDirection.none` | The direction the column is sorted                              |
+| sort-index     | `number` or `null` \*                                                                                        | `null`                          | The direction the column is sorted                              |
```
Comment (atmgrifter00) [inline packages/nimble-components/src/table/specs/table-column-sort-hld.md:39]: Looks like inadvertent copy-pasta of `sort-direction` description.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+| hidden         | `boolean`                                                                                                    | `false`                         | When set to true, do not render the column as part of the table |
+
+\* Note: The `sort-index` attribute is `number | null` because of the plan to use the `nullableNumberConverter` provided by FAST. That converter uses the value of `null` to represent non-number types rather than `undefined`, which is common within the nimble repo.
+
+### Validation
+
+The table will be invalid if it contains two columns with the same `sort-index`. If a duplicate `sort-index` is detected, the rows in the table will not render, no sort indicators will render, and the table's validity object will have `duplicateColumnSortIndex` set to `true`.
```
Comment (atmgrifter00) [inline packages/nimble-components/src/table/specs/table-column-sort-hld.md:46]: Is this level of rigidity necessary? I would think we treat identical sort indexes the same way any sort routine would treat identical values.

Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:46]: I could be convinced either way, but I tend to think that if we have a state that would lead to non-deterministic behavior, it should be invalid. Since having identical `sort-index` values would lead to non-deterministic behavior of what order we pass those columns to TanStack, I think we should consider it invalid. Though, the part of me that could be convinced otherwise, says that nothing will technically go wrong if the columns are sorted in a non-deterministic order.

I'd be curious to get the input of others -- @rajsite @jattasNI

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+### Visualizing the sort state
+
+Each column that is sorted will have an appropriate icon displayed in the header as shown in the design spec.
```
Comment (jattasNI) [inline packages/nimble-components/src/table/specs/table-column-sort-hld.md:50]: Do we have clarity on whether the icon will appear on all columns that contribute to the sort or just the primary one? I couldn't tell from a quick scan of the design spec.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md (hunk 1 above)
Comment (jattasNI) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:30]: I don't mind your proposal. But here are some ideas to spark discussion:
- `comparisonDataFieldName`
- `equalityDataFieldName`

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+## Implementation / Design
+
+### API
+
+The sorting state of the table will be declaritively defined within the `TableColumn` elements that are slotted within the table. A column will optionally be able to define its sort direction (i.e. `ascending` vs `descending`) and its sort index. The sort index enables a table to be sorted by multiple columns at the same time. For example, if a table should be sorted by a first name column and then a last name column, the first name column would have a sort index of 0, and the last name column would have a sort index of 1. This is a similar API to AG Grid, which provides [sort](https://www.ag-grid.com/javascript-data-grid/column-state/#reference-ColumnStateParams-sort) and [sortIndex](https://www.ag-grid.com/javascript-data-grid/column-state/#reference-ColumnStateParams-sortIndex) configurations.
+
+In some cases, a client may want to configure their table to be sorted by a column that is not visible within the table. To enable this, a column within the table can be marked as hidden. A hidden column exists in the table to provide metadata that can be used for operations, such as sorting, but it is not rendered within the table.
```
[the remark this answers is not in the document]
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:28]: It's not in SLE at this point in time, but it's in Leslie's wireframes. The grid configuration slide out should have 3 tabs (currently there are only 2). Those tabs are (1) visible columns, (2) sorted columns, (3) grouping. Given that long-term vision, I think it makes sense to allow sorting by non-visible columns rather than requiring clients to implement some workaround involving modifying the `operationalDataFieldName`. Similarly, it will be a requirement to group by non-visible columns (already a requirement in SLE), so I expect the concept of hidden columns will also be leveraged for that.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md (hunk 4 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:50]: Leslie's current guidance is that the icon will appear on all columns. I will clarify this sentence in the spec.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jattasNI) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:28]: Gotcha, I agree that use case would require explicit hidden columns.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+A column within the table may not have a one-to-one mapping to a field within the table's data. For example, a hyperlink column might use two fields from the data: a string to use as the hyperlink's display and a string to use as the hyperlink's href. Because of this, a column must define which data field is used when sorting that column. That field is specified by a column definition by implementing an `operationalDataFieldName` getter that returns the name of the field to use for sorting. For example, the text column would return the value of `fieldName` and the hyperlink column would return the field name associated with the display string.
+
+For performance reasons, the table will not support custom sort functions on a column. A column can only be sorted based on a single field within the table's data, which guards against performance degradations from inefficient custom sort functions.
```
[the remark this answers is not in the document]
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md]: That's a good question. TanStack has [a few different built-in sorting functions](https://tanstack.com/table/v8/docs/api/features/sorting#sorting-functions), but we can also provide our own within nimble (I'm not proposing changing our stance on client-specified sorting). None of TanStack's build-in sorting functions appear to be locale-aware. If you don't specify which sort function you want to use, TanStack will pick the one it thinks is most appropriate based on inspecting the data within the table. Looking at their code, I think we want to be more explicit with the way we handle sorting for all our columns.

I've updated the spec, but the summary of the new proposal is this:
Each table column (the column definition implementation, not exposed on the API of the column) specifies what sort is appropriate for it. At least until we find it necessary to add an override for `operationalDataFieldName`, the column should know the type associated with the field it's returning (e.g. text column would always return a field name with a field type of string, number column would always return a field name with a field type of number, etc). Therefore, a column definition will have a property on it that is the sort function to use, which returns a nimble-defined enum of the sorts we want to support. Within that implementation we will provide an option for locale-aware sort, that is backed by a private function we implement within nimble.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+For performance reasons, the table will not support custom sort functions on a column. Additionally, a column can only be sorted based on a single field within the table's data. These requirements guard against performance degradations from inefficient custom sort functions.
+
+#### Summary of new attributes on the base table column
```
Comment (rajsite) [inline packages/nimble-components/src/table/specs/table-column-sort-hld.md:36]: Can you describe more precisely the plugin author's api they need to implement?

Right now it sounds like they would implement a `get operationalDataFieldName()` and a `get sortFunction`.
I think to align with our `getColumnConfig` and `getDataRecordFieldNames` methods maybe this should instead be `getOperationalDataFieldName` and `getSortFunction`.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+In some cases, a client may want to configure their table to be sorted by a column that is not visible within the table. To enable this, a column within the table can be marked as hidden. A hidden column exists in the table to provide metadata that can be used for operations, such as sorting, but it is not rendered within the table.
+
+A column within the table may not have a one-to-one mapping to a field within the table's data. For example, a hyperlink column might use two fields from the data: a string to use as the hyperlink's display and a string to use as the hyperlink's href. Because of this, a column must define which data field is used when sorting that column. That field is specified by a column definition by implementing an `operationalDataFieldName` getter that returns the name of the field to use for sorting. For example, the text column would return the value of `fieldName` and the hyperlink column would return the field name associated with the display string.
```
Comment (rajsite) [inline packages/nimble-components/src/table/specs/table-column-sort-hld.md:30]: > For example, the text column would return the value of `fieldName` and the hyperlink column would return the field name associated with the display string.

I guess the getOperationalDataFieldName could exist in the  getDataRecordFieldNames but it doesn't need to right?

a hyperlink column could have a urlFieldName, displayFieldName, and valueFieldName where only the first two are used to render the cell template and the last one is used for operations like sorting / grouping, etc?

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md (hunk 7 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:36]: Maybe I'm mistaken because I haven't actually tried it out, but I'm pretty sure that we can't rely on the FAST's observables and the existing notification pattern if we make `getOperationalDataFieldName` and `getSortFunction` methods rather than properties. I envision us wanting to listen to changes to those values so we can re-run the sort.

FYI @atmgrifter00, I think you were planning on trying to remember/find the rationale behind making those other members functions rather than properties. Did you get a chance to do that?

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md (hunk 8 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:30]: That is correct. There is no requirement that the operationalDataFieldName is in the dataRecordFieldNames array.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md (hunk 7 above)
Comment (atmgrifter00) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:36]: @mollykreis I haven't but I will now go and try and do that.

Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:36]: The plan is to use properties for everything in the base class so that we ensure everything is observable. #1060

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+| attribute name | type                                                                                                         | default value                   | description                                                                                                                 |
+| -------------- | ------------------------------------------------------------------------------------------------------------ | ------------------------------- | --------------------------------------------------------------------------------------------------------------------------- |
+| sort-direction | `TableColumnSortDirection`, defined as `{none: undefined, ascending: 'ascending', descending: 'descending'}` | `TableColumnSortDirection.none` | The direction the column is sorted                                                                                          |
+| sort-index     | `number` or `null` \*                                                                                        | `null`                          | The index for sorting the column. When multiple columns are sorted, they will be sorted from lowest index to highest index. |
```
Comment (rajsite) [inline packages/nimble-components/src/table/specs/table-column-sort-hld.md:41]: > sorted from lowest index to highest index

Some edge cases we can test when assigned as attributes "-Infinity" "-2" "-0" "3.14" "1e6" "Infinity" "NaN"

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+The column definition will also be responsible for specifying the appropriate sort function to use based on the type of data it knows is associated with `operationalDataFieldName`. The sort function will be specified by a `sortFunction` getter on the column, that returns a `TableColumnSortFunction` value. `TableColumnSortFunction` is a new enum that will be introduced that defines the set of sort functions that can be applied to a column. Initially, the enum will consist of the value `basic`, which will correspond to [TanStack's `basic` sort function](https://tanstack.com/table/v8/docs/api/features/sorting#sorting-functions), and the value `localeAwareCaseSensitive`, which will be backed by a function written within nimble. That function will leverage [localeCompare()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare) to perform the sort.
```
[the remark this answers is not in the document]
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:32]: I like those names.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md (hunk 1 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:30]: While we haven't created specs for grouping or filtering, I do believe those operations would all be done using the same field (i.e. the one we're trying to name right now).

I'm good with `operandDataRecordFieldName`.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+#### Summary of new attributes on the base table column
+
+| attribute name | type                                                                                                         | default value                   | description                                                                                                                 |
+| -------------- | ------------------------------------------------------------------------------------------------------------ | ------------------------------- | --------------------------------------------------------------------------------------------------------------------------- |
+| sort-direction | `TableColumnSortDirection`, defined as `{none: undefined, ascending: 'ascending', descending: 'descending'}` | `TableColumnSortDirection.none` | The direction the column is sorted                                                                                          |
```
[the remark this answers is not in the document]
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:42]: I do think `none` is more descriptive/accurate in this case, so I'll update our docs. While I think people would figure it out without too much trouble, I think there might be confusion about what a `default` sort order means.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md (hunk 9 above)
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:43]: I made a note in the spec that we should test with these edge cases.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:36]: Done

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md]: - I agree with both your concern of relying on the host locale configuration and the thought that we could eventually configure the locale on the theme provider. I had actually thought about the same potential future work, and it's my mistake for not including that in this HLD.
- I think for now, it makes the most sense to not overly expose configuration options for the sorting behavior. This way, we can ensure everything is consistent and also only expose additional configuration options as required by valid use cases. 
- I'll do some investigation to see if we have standard backend behavior regarding the comparisons for grouping & sorting.

[the remark this answers is not in the document]
Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md]: - I've added the "future work" section to the HLD

- The understanding I've gained about backend comparisons is that nothing is locale-aware in the backend. The sorts/comparisons are case sensitive, but there's an ongoing discussion about whether the comparisons for filters should be case insensitive.

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+The column definition will also be responsible for specifying the appropriate sort function to use based on the type of data it knows is associated with `operandDataRecordFieldName`. The sort function will be specified by a `sortOperation` getter on the column, that returns a `TableColumnSortOperation` value. `TableColumnSortOperation` is a new enum that will be introduced that defines the set of sort functions that can be applied to a column. Initially, the enum will consist of the value `basic`, which will correspond to [TanStack's `basic` sort function](https://tanstack.com/table/v8/docs/api/features/sorting#sorting-functions), and the value `localeAwareCaseSensitive`, which will be backed by a function written within nimble. That function will leverage [localeCompare()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare) to perform the sort.
```
Comment (rajsite) [inline packages/nimble-components/src/table/specs/table-column-sort-hld.md:32]: > The sort function will be specified by a `sortOperation` getter on the column

Based on our recent API consistency changes I think this should be either an `abstract readonly` property or an `@observable` property for the column author apis.

**Review by rajsite — APPROVED** [review]

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (rajsite) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md]: Some of the backend investigation may be useful to capture in the future work section as well for when we consider what sortOperations to implement (which seems out of scope for this PR).

File: packages/nimble-components/src/table/specs/table-column-sort-hld.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
+# Table Column Sorting HLD
```
Comment (rajsite) [inline packages/nimble-components/src/table/specs/table-column-sort-hld.md:1]: Based on the backend discussions now I'm wondering if we should have considerations for a backend controlled sorting behavior?

For example, my data set is 1M elements, only the first 10K are sent to the UI. When I push the sort button in the header I don't want the table to sort the local data, I want to send the request to the server, get the response and put that into the table.

Is that a workflow we should / need to figure out right now?

Comment (rajsite) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:1]: Seems like we would want a cancellable sort event:
- User causes a sort action (i.e. changes sort direction in header for a column)
- Table fires an event that sort is changing
- Application code cancels that event to prevent the sort and fetches data from server with sort configuration
- Application loads data in table and sets the sort configuration (programmatically setting sort does not fire events, only user initiated, so we don't get stuck in a loop)
- Table runs sort on the current data and renders (which is already sorted but should be fast anyway)

Comment (mollykreis) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:1]: My initial thoughts are:
1. If all we're talking about is a cancellable event during interactive sort, then I don't think there is anything to be done now. That'll be covered as part of interactive sorting.
2. An interesting question is whether or not we want to have some mode (on a column? on the table?) that says don't actually sort the data, just mark it as sorted because the backend has already put it in the desired order.

I lean towards not complicating the API/code to try to handle (2) without having a solid understanding of the problem we're trying to solve (i.e. is performance an issue? is the sort algorithm the backend uses somehow different than the one the table is using & the rows end up in an unexpected/incorrect order? etc.). Therefore, I think it's fine for now to punt on the issue with the thinking that our 'sort' event will be cancellable. I'll add a note to the top of this HLD that this is something to consider when doing interactive sort.

Comment (rajsite) [reply packages/nimble-components/src/table/specs/table-column-sort-hld.md:1]: Makes sense to me 👍
